### ISSUE TITLE
Calculate energy of previously folded structure (specifically, dimer) in Python3

### ISSUE THREAD

**Issue by ihwang** [ViennaRNA/ViennaRNA#23]
Using the new Python3 API, what's the best way to calculate the free energy of a previously folded dimer? I've tried using `RNA.energy_of_structure(seq, struct, 0)` but that doesn't seem return the correct energy. Here's a simple example:
```
>>> RNA.fold_compound('CCC&GGG').mfe_dimer()
['((()))', -2.5]
>>> RNA.energy_of_structure('CCCGGG','((()))',0)
99993.3984375
````
Why is the energy of the simple dimer being returned as some exorbitantly large positive number?

**Comment by martin-mann** [comment]
Hi @ihwang 
I am no VRNA maintainer but think I can answer your question.

Eventually, you are comparing "apples and pears" :) Your first call is computing an *intermolecular* structure (dimer) while the second call is evaluating a single structure with *intramolecular* base pairs (no `&` in sequence nor structure; not sure if possible since not using tha python API). The energy of a dimer is differently computed even if its string encoding looks more or less the same.
```
E_dimer = E_init + E_hybridization = 4.1 + 3GCstackings

E_structure = E_hairpin + E_stacking = GChairpinwithoutfreebase + 3GCstackings
```
I am not even sure, if it is possible to compute an energy with a minimal loop size of 0 (no free bases enclosed by a base pair / hairpin). If so, this has to result in extremly high energies, since this is sterically quite impossible and thus should be very unfavorable in the structure model (= high energy).

To make things short: you have to call the energy evaluation for a dimer model instead of a single structure model! Since I am not familiar with the python API, I am not sure how to do. Maybe give it a try with added `&` symbols within the `energy_of_structure` call...

Best,
Martin

**Comment by martin-mann** [comment]
furthermore: `energy_of_structure` does *not* compute *minimum* free energy but just the free energy of a given structure. There is no optimization, since you have provided both sequence and structure. Just for nomenclature.. ;)

**Comment by ihwang** [comment]
Thank you for the clarification @martin-mann! As you can probably tell, my background is not in biology, and I appreciate the information and correction in nomenclature. I have corrected the title and original post of this issue.

I suspected this would be the case, since in [Issue 21](https://github.com/ViennaRNA/ViennaRNA/issues/21) I was informed that `.mfe_dimer()` had to be explicitly called in order to treat the sequence as two separate sequences being hybridized. I've tried putting `&` symbols in both the sequence and structure arguments of `energy_of_structure`, but this makes RNAfold throw an error.

Any further suggestions are appreciated! Perhaps main contributors of this repository can inform, like @RaumZeit? Thanks in advance!

**Comment by martin-mann** [comment]
when reading #21 : did you (reset) the cut_point when calling something like `energy_of_structure('CCC&GGG','(((&)))',0)`? maybe the underlying API uses the old datastructures... 

but as I said, I am no python API user and will shut up now.. ;)

**Comment by RaumZeit** [comment]
Hi ihwang,
evaluation of the free energy of dimers can be done in several ways. The is an old version of the API where one has to set to so-called cut_point position the defines where the second strand of the RNA actually starts in your input sequence. Since this is rather tedious, there are simplified wrappers of the evaluation functions, e.g. `RNA.eval_structure_simple()`.
I just tested it and found that the required input for dimers is a little bit strange, one has to add an ampersand into the sequence string separating the two strands, but the structure input needs to be free of the ampersand, e.g:
```
RNA.eval_structure_simple("CCC&GGG", "((()))")
```
works and retrieves the free energy of the 3 base pair helix of the two strands `CCC` and `GGG`.
As @martin-mann already pointed out, the free enery then not only consists of the stacking energies of the base pairs but also an initiation term for the dimer to form.
I hope this helps...

**Comment by ihwang** [comment]
Thanks @RaumZeit! Unfortunately when I try out `RNA.eval_structure_simple` I receive the error message `AttributeError: module 'RNA' has no attribute 'eval_structure_simple'`. I'm using ViennaRNA version 2.4.3.

**Comment by RaumZeit** [comment]
Hi @ihwang!
Sorry that you've been waiting so long for a fix. Indeed, version 2.4.3 did not provide the required energy evaluation functions and methods I mentioned above! Today, we've released version 2.4.4 which, however, provides all the required functions.
I hope this fixes the problems you've encountered.
